colossalai/shardformer/policies/llama.py

```diff
@@ -180,35 +195,24 @@ def module_policy(self):
 
         if self.pipeline_stage_manager:
             # set None as default
-            stage_manager = self.pipeline_stage_manager
-            layers_per_stage = Policy.distribute_layers(len(self.model.model.layers), stage_manager.num_stages)
-            stage_index = Policy.get_stage_index(layers_per_stage, stage_manager.stage)
-            method_replacement = {
-                'forward': partial(llama_for_causal_lm_forward, stage_manager=stage_manager, stage_index=stage_index)
-            }
-            self.append_or_create_method_replacement(description=method_replacement,
-                                                     policy=policy,
-                                                     target_key=LlamaForCausalLM)
+            self.set_pipeline_forward(model_cls=LlamaForCausalLM,
+                                      new_forward=LlamaPipelineForwards.llama_for_causal_lm_forward,
+                                      policy=policy)
+
         return policy
 
     def get_held_layers(self) -> List[Module]:
         """Get pipeline layers for current stage."""
-        module = self.model
         stage_manager = self.pipeline_stage_manager
-        held_layers = []
-        layers_per_stage = self.distribute_layers(len(module.model.layers), stage_manager.num_stages)
-        if stage_manager.is_first_stage():
-            held_layers.append(module.model.embed_tokens)
-        start_idx, end_idx = self.get_stage_index(layers_per_stage, stage_manager.stage)
-        held_layers.extend(module.model.layers[start_idx:end_idx])
+        held_layers = super().get_held_layers()
         if stage_manager.is_last_stage():
-            held_layers.append(module.model.norm)
-            held_layers.append(module.lm_head)
+            held_layers.append(self.model.lm_head)
         return held_layers
 
     def get_shared_params(self) -> List[Dict[int, Tensor]]:
         llama_model = self.model.model
-        if id(llama_model.embed_tokens.weight) == id(self.model.lm_head.weight):
+        if id(llama_model.embed_tokens.weight) == id(
```

Comment:
Here whether pipeline parallel has been enabled should be checked

Reply:
done